### ISSUE TITLE
possible problems with our list comprehension [x to y]

### ISSUE THREAD

**Issue by lutzhamel** [asteroid-lang/asteroid#181]
The problem with our list comprehension is that it never produces an empty list.  Consider,
```
> [1 to 1]
[1]
>
```
or
```
> [1 to 0]
[1,0]
> 
```
I think at least one of these cases needs to produce an empty list for the proper behavior of lists, especially in the context of for loops. Let's take a look at some use cases.  Iterating over lists, non-edge case,
```
> let l = ["zero","one"].
> for i in [0 to len(l)-1] do
.    io @println(l@i).
. end
zero
one
> 
```
That looks ok.  Next case, iterating over lists first edge case, list length of 1,
```
> load io.
> let l = ["zero"].
> for i in [0 to len(l)-1] do
.   io @println(l@i).
. end
zero
```
That also looks ok. Next case, trying to iterate over an empty list,
```
> load io.
> let l = [].
> for i in [0 to len(l)-1] do
.   io @println(l@i).
. end
error: list index out of range
>
```
Here is the problem!  The list comprehension computes a list of indexes that don't exist in the actual list,
```
> [0 to len(l)-1]
[0,-1]
>
```
I think the problem here is our "autodetect" of the step direction.  In the previous incarnation with explicit step directions like `step 1` or `step -1` the last case would have generated an empty list because there is not way to get from 0 to -1 with an implied default step of positive 1.

How do we fix this?  Any ideas?

**Comment by olwmc** [comment]
This may be the very reason python disallows implicated step values. I don't really see a simple way of getting around this. Hmm...

**Comment by lutzhamel** [comment]
Perhaps we need to reinstate the idea of a step value instead of a stride?

**Comment by olwmc** [comment]
That seems to make sense to me. I've been toying for a few minutes but the way we have it now doesn't really give any room to accommodate this case. I think it's reasonable to go back to step based list comprehensions

**Comment by lutzhamel** [comment]
I have another idea:  we change the behavior of the list comprehension to match the behavior of the `range` construction in Python. E.g.
```
Python 3.8.9 (default, Oct 26 2021, 07:25:53) 
[Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> list(range(0,2))
[0,1]
>>> list(range(0,1))
[0]
>>> list(range(0,0)) # this case and the next are semantically identical
[]
>>> list(range(0,0,-1))
[]
>>> list(range(0,-1,-1))
[0]
>>> 
```
except that we don't have to specify the explicit step value since it will be inferred by the list comprehension unless you would like a stride other than 1.  The Python examples written in Asteroid's list comprehension,
```
> [0 to 1]
[0]
> [0 to 0]
[]
> [0 to -1]
[0]
> [0 to -2]
[0,-1]
```
Then the problematic edge case from above would be,
```
load system io.
let l  = [].
for i in [0 to len(l)] do
  io @println(l@i).
end
```
and would print nothing since the list comprehension,
```
> [0 to len(l)]
[]
```
Notice that this approach also simplifies the canonical case of generating indexes for a particular list -- no pesky -1 needed on the list length.

**Comment by olwmc** [comment]
I like this!!! I can implement this today if you'd like :)

**Comment by lutzhamel** [comment]
Hold on. Just identified two more edge cases that will fail.  We have to go
back to the step notation ☹️

**Comment by olwmc** [comment]
Ahh brutal!

**Comment by olwmc** [comment]
Fixed with #187

**Comment by lutzhamel** [comment]
Sounds good.  I wonder how much in our test suite will break due to this
change 😒 Plus we need to update the documentation...
